Alicerce.playground/Pages/Network.xcplaygroundpage/Contents.swift

```diff
@@ -31,57 +33,170 @@ enum APIError: Error, Decodable {
-    let errorParser: (Data) -> APIError? = { try? JSONDecoder().decode(APIError.self, from: $0) }
+    typealias RetryMetadata = (request: Request, payload: External?, response: Response?)
 
-    let url: URL
-    let method: HTTP.Method
-    let headers: HTTP.Headers?
-    let query: HTTP.Query?
-    let body: Data?
+    typealias Error = GitHubAPIError
+    typealias ExternalMetadata = Response
 
-    init(url: URL,
-         method: HTTP.Method = .GET,
-         headers: HTTP.Headers? = nil,
-         query: HTTP.Query? = nil,
-         body: Data? = nil) {
+    // HTTPNetworkResource
 
-        self.url = url
-        self.method = method
-        self.headers = headers
-        self.query = query
-        self.body = body
+    let endpoint: Endpoint
+
+    // RetryableResource
+
+    var retryErrors: [Swift.Error] = []
+    var totalRetriedDelay: Retry.Delay = 0
+    let retryPolicies: [RetryPolicy]
+
+    // EmptyExternalResource
+
+    static var empty: External { return Data() }
+
+    // ExternalErrorDecoderResource
+
+    var decodeError: DecodeErrorClosure = { data, _ in
+        
+        guard let data = data else { return nil }
+        return try? JSONDecoder().decode(Error.self, from: data)
+    }
+
+    init(endpoint: Endpoint, retryPolicies: [RetryPolicy] = []) {
+
+        self.endpoint = endpoint
+        self.retryPolicies = retryPolicies
     }
 }
 
 // Model
 
-struct Model: Codable {
-    // ...
+struct GitHubRepo: Decodable {
+    var name: String
+    var fullName: String
+    var stars: Int
+
+    private enum CodingKeys: String, CodingKey {
+        case name
+        case fullName = "full_name"
+        case stars = "stargazers_count"
+    }
 }
 
+struct Dummy: Decodable {}
+
 // Request
 
-let resource = RESTResource<Model>(url: URL(string: "http://localhost/")!)
+let repoResource = GitHubResource<GitHubRepo>(endpoint: .repo(owner: "Mindera", name: "Alicerce"))
+let nonExistentResource = GitHubResource<Dummy>(endpoint: .nonExistent)
+
+network.fetch(resource: repoResource) { result in
+
+    switch result {
+    case .success(let value):
+        String(bytes: value.value, encoding: .utf8)
+    case .failure(.api(let apiError as GitHubAPIError, let statusCode, let response)):
+        apiError
+        statusCode
+        response
+    case .failure(let error):
+        error
+    }
+}
+
+network.fetch(resource: nonExistentResource) { result in
+
+    switch result {
+    case .success(let value):
+        String(bytes: value.value, encoding: .utf8)
+    case .failure(.api(let apiError as GitHubAPIError, let statusCode, let response)):
+        apiError
+        statusCode
+        response
+    case .failure(let error):
+        error
+    }
+}
+
+// NetworkStore (via NetworkStack)
+
+extension GitHubResource: DecodableResource & PersistableResource & NetworkStoreStrategyFetchResource {
+
+    // DecodableResource
+
+    var decode: DecodeClosure { return { try JSONDecoder().decode(T.self, from: $0) } }
```

Comment:
This could be the default implementation for a `protocol JSONDecodableResource: DecodableResource ` where `Internal: Decodable` and `External == Data`

Reply:
I thought about that, but actually we can't because we can't assume the encoding is JSON whenever the `External` type is `Data` 😜

Reply:
Of course, that's why I suggested a new protocol to wrap that up. Something along the lines of:
```
protocol JSONDecodableResource: DecodableResource {}

extension JSONDecodableResource where Internal: Decodable, Extenal == Data {

    var decode: DecodeClosure { return { try JSONDecoder().decode(Internal.self, from: $0) } }
}
```

Reply:
Oops, missed that sorry 🙈 It would be more correct, indeed.

I'm personally not a big fan of instantiating a new decoder on each invocation, but we can easily work around this and even make it more generic by creating a `decoder: Decoder` property on the protocol. Then it would be assumed that the decoder you choose will be the right one to whatever format you are expecting.